python_basics_writing_python_code/python_basics_writing_python_code.md

```diff
@@ -0,0 +1,923 @@
+    </div>
+    ***
+
+## Dictionaries
+
+Another kind of collection object in Python is a **dictionary**. Dictionaries are similar to lists in some ways, in that they are ordered (in recent versions of Python; in older versions, dictionaries are unordered) and changeable. However, there are some important distinctions:
+
+* Dictionaries are collections of **key-value pairs**. For example if you had a dictionary of demographic information for an individual, the **keys** might be the type of information stored (address, telephone number, etc.) and the **values** would be the actual values of those data (123 Puppydog Lane, 987-654-3210).
+
+* They do not allow duplicate key-value pairs.
+
+* Dictionaries use curly brackets { }.
+
+Let's build an example dictionary:
+
+@sage
+<div class="python_link">
+<lia-keep>
+<script type="text/x-sage">
+
+contact_info = {"address" : "123 Puppydog Lane",
+                "telephone" : "987-654-3210",
+                "email" : "myname@email.com"}
+print(contact_info)
+
+</script>
+</lia-keep>
+</div>
+
+Like lists, dictionaries have some useful methods you can use to access or alter their data:
+
+* `.keys()` returns a list of the dictionary's keys.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.keys())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.values()` returns a list of the dictionary's values.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.values())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.items()` returns a list of the dictionary's key-value pairs as tuples (an object that holds multiple values, similar to a list, except that they are unchangeable). This may seem not very useful at first glance, but since different collection objects have different properties, there could be circumstances in which it is more useful to have the data in a different kind of object.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.items())
+
+    </script>
+    </lia-keep>
+    </div>
+
+* `.pop()` works similarly for dictionaries as it does for lists, except that it removes an element by its key.
+
+    @sage
+    <div class="python_link">
+    <lia-keep>
+    <script type="text/x-sage">
+
+    print(contact_info.pop("email"))
+print(contact_info)
+
+    </script>
+    </lia-keep>
+    </div>
+
+    <div class="warning">
+    Note that this change is permanent! If you want to keep the original dictionary as-is, you should make a copy of the dictionary by assigning it to a new variable, and then using `.pop()` on the copy. In this example though, all you need to do to restore the dictionary to its original form is re-run the first code cell or refresh the page.
```

Comment:
```suggestion
    Note that this change is permanent! If you want to keep the original dictionary as-is, you should make a copy of the dictionary (using the `.copy()` method and assigning the copy to a new variable), and then using `.pop()` on the copy. In this example though, all you need to do to restore the dictionary to its original form is re-run the first code cell or refresh the page.
```

I haven't tested this so doublecheck that `.copy()` works the way I think, or include an example of how to do this just below this section